src/interfaces/records/records-interface.ts

```diff
@@ -1,10 +1,56 @@
-import { MessageStore } from '../../store/message-store.js';
-import { RecordsWrite } from './messages/records-write.js';
-import { RecordsWriteMessage } from '../../interfaces/records/types.js';
-import { TimestampedMessage } from '../../core/types.js';
+import type { DataStore } from '../../store/data-store.js';
+import type { MessageStore } from '../../store/message-store.js';
+import type { RecordsWriteMessage } from '../../interfaces/records/types.js';
+import type { TimestampedMessage } from '../../core/types.js';
 
 import { constructRecordsWriteIndexes } from './handlers/records-write.js';
 import { Message } from '../../core/message.js';
+import { RecordsWrite } from './messages/records-write.js';
+
+/**
+ * Deletes all data for messages in `existingMessages` that are older than the `comparedToMessage` in the given tenant,
+ * but only if they have the same recordId and a different dataCid (i.e. same message with different data).
+ */
+export async function deleteAllOlderData(
+  tenant: string,
+  existingMessages: TimestampedMessage[],
+  comparedToMessage: TimestampedMessage,
+  dataStore: DataStore
+): Promise<void> {
+  const recordId = Message.getRecordId(comparedToMessage);
+
+  const dataCidToDelete = new Set<string>();
+
+  // NOTE: under normal operation, there should only be at most two existing records per `recordId` (initial + a potential subsequent write/delete),
+  // but the DWN may crash before `delete()` is called below, so we use a loop as a tactic to clean up lingering data as needed
+  for (const message of existingMessages) {
+    const messageIsOld = await Message.isOlder(message, comparedToMessage);
+    if (!messageIsOld) {
+      continue;
+    }
+
+    // ignore anything that's not the same record
+    if (Message.getRecordId(message) !== recordId) {
+      continue;
+    }
```

Comment:
Consider either:

- remove this check as this code path can't be hit since all messages given will be in the scope of the same `recordId` already; or
- if you intended for this method to be more general purpose, add a test to cover the path